sql/handler.cc

```diff
@@ -456,15 +456,9 @@ int ha_finalize_handlerton(st_plugin_int *plugin)
   if (!hton)
     goto end;
 
-  switch (hton->state) {
-  case SHOW_OPTION_NO:
-  case SHOW_OPTION_DISABLED:
-    break;
-  case SHOW_OPTION_YES:
-    if (installed_htons[hton->db_type] == hton)
-      installed_htons[hton->db_type]= NULL;
-    break;
-  };
+  if (ha_storage_engine_is_enabled(hton) &&
+      installed_htons[hton->db_type] == hton)
+    installed_htons[hton->db_type]= NULL;
```

Comment:
`installed_htons` are mostly used by `ha_resolve_by_legacy_type()`. It is intended to open old tables (previously table metadata contained storage engine id, e.g. 12; now it contains storage engine name, e.g. "InnoDB").

Since one couldn't create "binlog" tables anyway, it looks like you could avoid calling `ha_storage_engine_is_enabled()` by removing `binlog_hton->db_type=DB_TYPE_BINLOG`.